kale/utils/download.py

```diff
@@ -0,0 +1,71 @@
+# =============================================================================
+# Author: Xianyuan Liu, xianyuan.liu@sheffield.ac.uk
+#         Raivo Koot, rekoot1@sheffield.ac.uk
+#         Haiping Lu, h.lu@sheffield.ac.uk or hplu@ieee.org
+# =============================================================================
+
+"""Data downloading and compressed data extraction functions, Based on
+https://github.com/pytorch/vision/blob/master/torchvision/datasets/utils.py
+https://github.com/pytorch/pytorch/blob/master/torch/hub.py
```

Comment:
I assume that you have followed our 3R green ML principles here to reuse pytorch APIs.

Reply:
However, it should be clear why the original APIs in pytorch is not good/simple enough. Why the user should use this API rather than the pytorch ones directly. What are the added benefits of using this API.

Reply:
1. `download_and_extract_archive` and `download_url_to_file` are not in the same PyTorch file and It would be better to integrate them into one file. This file cover functions to download files with different types from different sources. Using the same format and parameters like `output_directory` and `output_file_name` can also simplify our coding.

2. PyTorch provides the basic download function but it may raise issues when the `output_directory` does not exist. I only meet this problem.